source/common/quic/envoy_quic_stream.cc

```diff
@@ -102,10 +102,89 @@ void EnvoyQuicStream::encodeTrailersImpl(spdy::Http2HeaderBlock&& trailers) {
+  memcpy(&data[0], str.data(), str.length()); // NOLINT(safe-memcpy)
+  return data;
+}
+
+void serializeMetadata(const Http::MetadataMapPtr& metadata, quic::QuicStreamId id,
+                       absl::InlinedVector<quiche::QuicheMemSlice, 2>& slices) {
+  quic::NoopDecoderStreamErrorDelegate decoder_stream_error_delegate;
+  quic::QpackEncoder qpack_encoder(&decoder_stream_error_delegate,
+                                   quic::HuffmanEncoding::kDisabled);
+
+  spdy::Http2HeaderBlock header_block;
+  for (const auto& [key, value] : *metadata) {
+    header_block.AppendValueOrAddHeader(key, value);
+  }
+
+  // The METADATA frame consist of a frame header, which includes payload
+  // length, and a payload, which is the QPACK-encoded metadata block. In order
+  // to generate the frame header, the payload needs to be generated first.
+  std::string metadata_frame_payload =
+      qpack_encoder.EncodeHeaderList(id, header_block,
+                                     /* encoder_stream_sent_byte_count = */ nullptr);
+  std::string metadata_frame_header =
+      quic::HttpEncoder::SerializeMetadataFrameHeader(metadata_frame_payload.size());
+
+  slices.emplace_back(dataFromString(metadata_frame_header), metadata_frame_header.length());
+  slices.emplace_back(dataFromString(metadata_frame_payload), metadata_frame_payload.length());
+}
+
+} // namespace
+
+void EnvoyQuicStream::encodeMetadata(const Http::MetadataMapVector& metadata_map_vector) {
+  if (!http3_options_.allow_metadata()) {
+    // Metadata Frame is not supported in QUICHE.
+    ENVOY_STREAM_LOG(debug, "METADATA is not supported in Http3.", *this);
+    stats_.metadata_not_supported_error_.inc();
+    return;
+  }
+  if (quic_stream_.write_side_closed()) {
+    IS_ENVOY_BUG("encodeMetadata is called on write-closed stream.");
+    return;
+  }
+  ASSERT(!local_end_stream_);
+
+  for (const Http::MetadataMapPtr& metadata : metadata_map_vector) {
+    absl::InlinedVector<quiche::QuicheMemSlice, 2> quic_slices;
+    quic_slices.reserve(2);
+    serializeMetadata(metadata, quic_stream_.id(), quic_slices);
+    absl::Span<quiche::QuicheMemSlice> metadata_frame(quic_slices);
+
+    SendBufferMonitor::ScopedWatermarkBufferUpdater updater(&quic_stream_, this);
+    quic::QuicConsumedData result{0, false};
+    {
+      IncrementalBytesSentTracker tracker(quic_stream_, *mutableBytesMeter(), false);
+      result = quic_stream_.WriteMemSlices(metadata_frame, /*end_stream=*/false);
+    }
+    // QUIC stream must take all.
+    if (result.bytes_consumed == 0) {
+      IS_ENVOY_BUG(fmt::format("Send buffer didn't take all the data. Stream is write {} with {} "
+                               "bytes in send buffer. Current write was rejected.",
+                               quic_stream_.write_side_closed() ? "closed" : "open",
+                               quic_stream_.BufferedDataBytes()));
+      quic_stream_.Reset(quic::QUIC_BAD_APPLICATION_PAYLOAD);
```

Comment:
is bad application payload the best?  feels like an invariant failure but idk if ther's a better code.